### ISSUE TITLE
Review the 'groups_which_view_metadata'

### ISSUE THREAD

**Issue by darrenleeweber** [sul-dlss/dor-services#146]
With regard to:
```
def can_view_metadata?(roles)
  intersect roles, groups_which_view_metadata
end
```

The `groups_which_view_metadata` is a misnomer, it should be `roles_which_view_metadata`.

With regard to related issues in argo, like https://github.com/sul-dlss/argo/issues/203 and https://github.com/sul-dlss/argo/issues/76, there may be users who belong to groups that have a role in the `dor-apo-viewer`, but this role will never be able to view an item because it is not included in the `roles_which_view_metadata`.  For example, in a spec of argo, we can see:
```
[11] pry(#<RSpec::ExampleGroups::Apo::WithViewingUser>)> dorObj.groups_which_view_metadata
=> ["dor-administrator", "sdr-administrator", "dor-apo-manager", "dor-apo-depositor", "dor-viewer", "sdr-viewer"]
[12] pry(#<RSpec::ExampleGroups::Apo::WithViewingUser>)> dorObj.roles
=> {"dor-apo-manager"=>["workgroup:dlss:developers"], "dor-apo-viewer"=>["person:sunetid:someone"]}
```

For this DOR APO object, the DOR object has a set of `groups_which_view_metadata` and a set of `roles`.  If the `dor-apo-viewer` role is matched for a user (based on the user workgroups), it doesn’t allow viewing the APO because there is no `dor-apo-viewer` in the `groups_which_view_metadata`.

So, for the APO `dorObj.groups_which_view_metadata`, should there be a `dor-apo-viewer`  in that set?

**Comment by darrenleeweber** [comment]
Similarly,
```
> d = Dor::AdminPolicyObject.find(@druid)
=> #<Dor::AdminPolicyObject pid:"druid:ab987mn1234", >
> d.is_a? Dor::AdminPolicyObject
=> true
> d.can_view_content? ["dor-viewer"]
=> true
> d.can_view_content? ["dor-apo-viewer"]
=> false
> d.can_view_metadata? ["sdr-viewer"]
=> true
```

**Comment by darrenleeweber** [comment]
This is a blocker for https://github.com/sul-dlss/argo/issues/76, where @LynnMcRae says the role definitions need to be updated, see https://github.com/sul-dlss/argo/issues/76#issuecomment-171129115

**Comment by darrenleeweber** [comment]
@LynnMcRae -- please review https://github.com/sul-dlss/dor-services/blob/develop/lib/dor/models/governable.rb#L78-L101 and advise what all these should be?

**Comment by atz** [comment]
My understanding is that there not a `dor-apo-viewer` is because the APO is already also a modeled object.  Instead of asking:
```
i = Dor::Item.find(@druid1)
i.can_view_content? ["dor-apo-viewer"]
```
the question should be the same as for any other object, as you attempted:
```
apo = Dor::AdminPolicyObject.find(@druid2)
apo.can_view_content? ["dor-viewer"]
```

The permissions linkage you are trying to make is not a question of what broad groups of users have access to broad types of objects.  Dor answers that correctly.  You are trying to ask whether *this particular user* can access *this particular APO*.  `dor-services` is interested in answering that exact question only inasmuch as it indexes (and here provides convenience methods for interrogating) permissions groups/roles/strings: it has no notion of an application's User class.  

Argo can handle this itself by modifying what `roles` returns, **if** the correctly solrized fields are available from the APO object.  Out of 886595 docs in the `argo_b` index, only [16](https://sul-solr.stanford.edu/solr/#/argo_b_shard1_replica3/schema-browser?field=apo_role_group_dor-apo-viewer_ssim) have **apo_role_dor-apo-viewer_ssim** values indexed.  Note there is also:

 - apo_role_person_dor-apo-viewer_ssim
 - apo_role_group_dor-apo-viewer_ssim
 - apo_role_hydrus-collection-viewer_ssim
 - etc

All "apo_role_*" index values are derived from forms in [RoleMetadataDS](https://github.com/sul-dlss/dor-services/blob/develop/lib/dor/datastreams/role_metadata_ds.rb) (because only APOs have a RoleMetadataDS).  So that is where our ugly data is hiding.

TL;DR: IMHO, we should be asking `apo.can_view_content? ["dor-viewer"]`, not `x.can_view_content? ["dor-apo-viewer"]`

**Comment by darrenleeweber** [comment]
Note there is a lot of duplication in the governable groups (aka roles) and the methods should be constants.  That is, this is the current state of affairs:
```
    def groups_which_manage_item
      ['dor-administrator', 'sdr-administrator', 'dor-apo-manager', 'dor-apo-depositor']
    end
    def groups_which_manage_desc_metadata
      ['dor-administrator', 'sdr-administrator', 'dor-apo-manager', 'dor-apo-depositor', 'dor-apo-metadata']
    end
    def groups_which_manage_system_metadata
      ['dor-administrator', 'sdr-administrator', 'dor-apo-manager', 'dor-apo-depositor']
    end
    def groups_which_manage_content
      ['dor-administrator', 'sdr-administrator', 'dor-apo-manager', 'dor-apo-depositor']
    end
    def groups_which_manage_rights
      ['dor-administrator', 'sdr-administrator', 'dor-apo-manager', 'dor-apo-depositor']
    end
    def groups_which_manage_embargo
      ['dor-administrator', 'sdr-administrator', 'dor-apo-manager', 'dor-apo-depositor']
    end
    def groups_which_view_content
      ['dor-administrator', 'sdr-administrator', 'dor-apo-manager', 'dor-apo-depositor', 'dor-viewer', 'sdr-viewer']
    end
    def groups_which_view_metadata
      ['dor-administrator', 'sdr-administrator', 'dor-apo-manager', 'dor-apo-depositor', 'dor-viewer', 'sdr-viewer']
    end
```

This can all be replaced with a combination of two or three constants, something like:
```
    ADMIN_ROLES =  ['dor-administrator', 'sdr-administrator', 'dor-apo-manager', 'dor-apo-depositor']
    VIEWER_ROLES = ['dor-viewer', 'sdr-viewer']
    def admins
      ADMIN_ROLES
    end
    alias_method :groups_which_manage_item, :admins
    alias_method :groups_which_manage_system_metadata, :admins
    alias_method :groups_which_manage_content, :admins
    alias_method :groups_which_manage_rights, :admins
    alias_method :groups_which_manage_embargo, :admins
    def groups_which_manage_desc_metadata
      ADMIN_ROLES + ['dor-apo-metadata']
    end
    def viewers
      ADMIN_ROLES + VIEWER_ROLES
    end
    alias_method :groups_which_view_content, :viewers
    alias_method :groups_which_view_metadata, :viewers
```

**Comment by darrenleeweber** [comment]
@LynnMcRae says:



Concerning which-roles exist per above, first I think there is some vocabulary updating needed for this list -- "dor-administrator", "sdr-administrator", "dor-apo-manager", "dor-apo-depositor", "dor-viewer"

• dor-administrator is the old name for sdr-administrator and can be dropped
• there should be an sdr-viewer for global viewing rights, replacing dor-viewer
• I don't think dor-apo-depositor has an implementation and we don't use it, so it can be dropped.
• there should indeed be a dor-apo-viewer

Other roles are postulated but not yet implemented. I've made that clearer on the consul page.

This would make the current list

"sdr-administrator", "sdr-viewer", "dor-apo-manager", "dor-apo-viewer"

supporting read-only vs read/write access both globally and per-APO.

**Comment by darrenleeweber** [comment]
https://github.com/sul-dlss/dor-services/pull/148 does not address the role definition changes noted by @LynnMcRae above.  It maintains the previous state of affairs, i.e.
```
    ROLES4ADMINS =  %w(
      dor-administrator sdr-administrator
      dor-apo-manager dor-apo-depositor
    ).freeze

    ROLES4VIEWERS = %w(dor-viewer sdr-viewer).freeze
```
The permitted viewers are `ROLES4ADMINS + ROLES4VIEWERS`.  The one addition to these lists is for managers of metadata, which is `ROLES4ADMINS + ['dor-apo-metadata']`.  If Lynn's revisions are applies, the changes seem to be:
```
    ROLES4ADMINS =  %w(sdr-administrator dor-apo-manager).freeze
    ROLES4VIEWERS = %w(sdr-viewer dor-apo-viewer).freeze
```
